ui/ops/src/dynamic/widgets/usage/usage.js

```diff
@@ -0,0 +1,192 @@
+      {} // keyed by the leading edges .getTime()
+  );
+  asyncWatch([() => toValue(name), () => toValue(edges)], async ([name, edges], [oldName]) => {
+    if (name !== oldName) {
+      Object.keys(countsByEdge).forEach(k => delete countsByEdge[k]);
+    }
+
+    const toDelete = new Set(Object.keys(countsByEdge));
+    for (const edge of edges) {
+      toDelete.delete(edge.getTime().toString());
+    }
+    for (const k of toDelete) {
+      delete countsByEdge[k];
+    }
+
+    for (let i = 0; i < edges.length - 1; i++) {
+      const leadingEdge = edges[i];
+      if (!countsByEdge[leadingEdge.getTime()]) {
+        const toFetch = [leadingEdge];
+        for (let j = i + 1; j < edges.length; j++) {
+          const e = edges[j];
+          toFetch.push(e);
+          if (countsByEdge[e.getTime()]) {
+            i = j - 1; // will i++ in the loop
+            break;
+          }
+        }
+
+        const countBefore = countsByEdge[edges[i - 1]?.getTime()]?.last ?? null;
+        const records = await readUsageCountSeries(name, toFetch, countBefore);
+        for (const dataset in records) {
+          const groupData = records[dataset];
+          for (const record of groupData) {
+            if (!countsByEdge[record.x.getTime()])
+              countsByEdge[record.x.getTime()] = {};
+
+            countsByEdge[record.x.getTime()][dataset] = record.y;
+          }
+        }
+      }
+    }
+  }, {immediate: true});
+
+  return computed(() => {
+    return Object.entries(countsByEdge)
+        .map(([timestamp, groupData]) => ({
+          x: new Date(Number(timestamp)),
+          groups: groupData
+        }))
+        .sort((a, b) => a.x.getTime() - b.x.getTime());
+  });
+}
+
+/**
+ * Reads and calculates the total usage count for each span between the edges, returning it as a data series.
+ * The y property at index i in the response will be the total usage count for the period between edges[i] and edges[i+1].
+ * The last property at index i will be the most recent recorded usage count for the period between edges[i] and edges[i+1].
+ *
+ * @param {string} name
+ * @param {Date[]} edges
+ * @param {number | null} [countBefore] - the usage count before the first edge, if not null
+ * @return {Promise<{string: {x: Date, y: number, last: number}[]}>} - of size edges.length - 1
+ */
+async function readUsageCountSeries(name, edges, countBefore) {
+  const findEdges = (edges, at) => {
+    let i = binarySearch(edges, at, (a, b) => a.getTime() - b.getTime());
+    if (i < 0) {
+      // binarySearch will return the index _after_ the edge before the span as this is where the value would be inserted
+      i = ~i - 1;
+    }
+    const res = {beforeIdx: i, before: edges[i], after: null, afterIdx: null};
+    if (i < edges.length - 1) {
+      res.after = edges[i + 1];
+      res.afterIdx = i + 1;
+    }
+    return res;
+  }
+
+  /** @type {({x: Date, y: number, last: number}[])} */
+  const dates = Array(edges.length - 1).fill(null);
```

Comment:
I'm struggling to know what this is from the name, it certainly isn't containing any dates.